lisa/tools/hugepages.py

```diff
@@ -0,0 +1,115 @@
+        and extracting the size from the directory name.
+        e.g. [
+        "'/sys/devices/system/node/node0/hugepages/hugepages-16777216kB/'",
+        "'/sys/devices/system/node/node0/hugepages/hugepages-2048kB/'",
+        "'/sys/devices/system/node/node0/hugepages/hugepages-524288kB/'"
+        ]
+        """
+        ls = self.node.tools[Ls]
+        # if not ls.path_exists(self._hugepages_dir_path, sudo=True):
+        #     raise SkippedException(f"path: {self._hugepages_dir_path} does not exist")
+        hugepage_size_dir_names = ls.list_dir(self._hugepages_dir_path, sudo=True)
+        for hugepage_size_dir_name in hugepage_size_dir_names:
+            matched_hugepage = PATTERN_HUGEPAGE.match(hugepage_size_dir_name)
+            if not matched_hugepage:
+                raise LisaException(
+                    f"no hugepage size found in {hugepage_size_dir_name}"
+                )
+            self._hugepage_sizes.add(int(matched_hugepage.group("hugepage_size_in_kb")))
+
+    def _enable_hugepages(
+        self, hugepage_size_kb: HugePageSize = HugePageSize.HUGE_2MB
+    ) -> None:
+        echo = self.node.tools[Echo]
+        meminfo = self.node.tools[Free]
+        nics_count = len(self.node.nics.get_nic_names())
+        numa_nodes = self.node.tools[Lscpu].get_numa_node_count()
+
+        request_space_kb = (nics_count - 1) * 1024 * 1024 * numa_nodes * 2
+        free_memory_kb = meminfo.get_free_memory_kb()
+
+        if free_memory_kb < request_space_kb:
+            self.node.log.debug(
+                f"WARNING: Not enough {hugepage_size_kb.value} KB pages "
+                "available for DPDK! "
+                f"Requesting {request_space_kb} KB found {free_memory_kb} "
+                "KB free."
+                "Test may fail if it cannot allocate memory."
+            )
+        request_pages = request_space_kb // hugepage_size_kb.value
+        for i in range(numa_nodes):
+            # nr_hugepages will be written with the number calculated
+            # based on 2MB hugepages if not specified, subject to change
+            # this based on further discussion
+            echo.write_to_file(
+                f"{request_pages}",
+                self.node.get_pure_path(
+                    f"/sys/devices/system/node/node{i}/hugepages/"
+                    f"hugepages-{hugepage_size_kb.value}kB/nr_hugepages"
+                ),
+                sudo=True,
+            )
+
+    def init_hugepages(self, hugepage_size: Optional[HugePageSize] = None) -> None:
+        mount = self.node.tools[Mount]
+        if hugepage_size:
+            self.get_hugepage_sizes_in_kb()
+            if hugepage_size.value not in self._hugepage_sizes:
+                raise SkippedException(
+                    f"Supported hugepage sizes in kB are: {self._hugepage_sizes}."
+                    f" {hugepage_size.value} kB size is not available"
+                )
+            mount.mount(
```

Comment:
I'm not sure if this is needed on all platforms. You could try checking if sysfs already has the file available for enabling the hugepage size first?

My memory is that this was only a 'don't care' function call before because it was needed on old CentOS versions or something and I was being lazy. Now that we're rewriting it; you can code it better than I did 😜

Reply:
Added the step to check if it is mounted already